tensornetwork.py

```diff
@@ -902,14 +917,38 @@ def contract_between(self,
       ValueError: If no edges are found between node1 and node2 and
         `allow_outer_product` is set to `False`.
     """
-    flat_edge = self.flatten_edges_between(node1, node2)
-    if not flat_edge:
+    # Trace edges cannot be contracted using tensordot.
+    if node1 is node2:
+      flat_edge = self.flatten_edges_between(node1, node2)
+      if flat_edge:
+        return self.contract(flat_edge, name)
+      else:
+        raise ValueError("No trace edges found on contraction of edges between "
+                         "node '{}' and itself.".format(node1))
+
+    shared_edges = self.get_shared_edges(node1, node2)
+    if not shared_edges:
       if allow_outer_product:
         return self.outer_product(node1, node2)
       else:
         raise ValueError("No edges found between nodes '{}' and '{}' "
                          "and allow_outer_product=False.".format(node1, node2))
-    return self.contract(flat_edge, name)
+
+    axes1 = []
+    axes2 = []
+    for edge in shared_edges:
+      if edge.node1 is node1:
+        axes1.append(edge.axis1)
```

Comment:
Add a comment explaining why this if else is needed. (I know why, but it might not be obvious for new people).